Kernel/Net/RTL8168NetworkAdapter.cpp

```diff
@@ -0,0 +1,384 @@
+ * Copyright (c) 2021, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/MACAddress.h>
+#include <Kernel/Debug.h>
+#include <Kernel/IO.h>
+#include <Kernel/Net/RTL8168NetworkAdapter.h>
+
+namespace Kernel {
+
+#define REG_MAC 0x00
+#define REG_MAR4 0x0B
+#define REG_MAR0 0x0F
+#define REG_TXADDR 0x20
+#define REG_COMMAND 0x37
+#define REG_IMR 0x3C
+#define REG_ISR 0x3E
+#define REG_TXCFG 0x40
+#define REG_RXCFG 0x44
+#define REG_CFG9346 0x50
+#define REG_PHYACCESS 0x60
+#define REG_PHYSTATUS 0x6C
+#define REG_RMS 0xDA
+#define REG_CPLUS_COMMAND 0xE0
+#define REG_RXADDR 0xE4
+#define REG_MTPS 0xEC
+
+#define COMMAND_TX_ENABLE 0x04
+#define COMMAND_RX_ENABLE 0x08
+#define COMMAND_RESET 0x10
+
+#define CPLUS_COMMAND_VLAN_DETAG 0x40
+
+#define INT_RXOK 0x01
+#define INT_RXERR 0x02
+#define INT_TXOK 0x04
+#define INT_TXERR 0x08
+#define INT_RX_DESCRIPTOR_UNAVAILABLE 0x10
+#define INT_LINK_CHANGE 0x20
+#define INT_RX_FIFO_OVERFLOW 0x40
+
+#define CFG9346_NONE 0x00
+#define CFG9346_EEM0 0x40
+#define CFG9346_EEM1 0x80
+
+#define TXCFG_MAX_DMA_1K 0x700
+#define TXCFG_IFG011 0x1000000
+
+#define RXCFG_APM 0x02
+#define RXCFG_AM 0x04
+#define RXCFG_AB 0x08
+#define RXCFG_MAX_DMA_UNLIMITED 0x0700
+#define RXCFG_FTH_NONE 0xE000
+
+#define PHY_LINK_STATUS 0x02
+
+#define PHY_FLAG 0x80
+#define PHY_REG_BMCR 0x0
+
+#define BMCR_SPEED_0 0x2000
+#define BMCR_AUTO_NEGOTIATE 0x1000
+#define BMCR_DUPLEX 0x100
+#define BMCR_SPEED_1 0x40
+
+#define MTPS_JUMBO 0x3F
+
+#define PACKET_BUFFER_SIZE 0x1FF8 // Maximum allowed by RX descriptor, and a bit larger than needed for TX descriptor
```

Comment:
This causes an assertion failure since it's not page aligned:
```
[34;1m[init_stage2(2:2)][0m: ASSERTION FAILED: !(size % PAGE_SIZE)
[34;1m[init_stage2(2:2)][0m: ../../Kernel/VM/MemoryManager.cpp:486 in AK::OwnPtr<Kernel::Region> Kernel::MemoryManager::allocate_contiguous_kernel_region(size_t, AK::String, Kernel::Region::Access, size_t, Kernel::Region::Cacheable)
[34;1m[#0 init_stage2(2:2)][0m: 0xc05437a5 (next: 0xc1135b70)
[34;1m[#0 init_stage2(2:2)][0m: 0xc04e9451 (next: 0xc1135be0)
[34;1m[#0 init_stage2(2:2)][0m: 0xc0617046 (next: 0xc1135c60)
[34;1m[#0 init_stage2(2:2)][0m: 0xc06189fa (next: 0xc1135d20)
[34;1m[#0 init_stage2(2:2)][0m: 0xc031b9f9 (next: 0xc1135d90)
[34;1m[#0 init_stage2(2:2)][0m: 0xc034c75d (next: 0xc1135de0)
[34;1m[#0 init_stage2(2:2)][0m: 0xc034b235 (next: 0xc1135e60)
[34;1m[#0 init_stage2(2:2)][0m: 0xc034b297 (next: 0xc1135e80)
[34;1m[#0 init_stage2(2:2)][0m: 0xc061684c (next: 0xc1135ed0)
[34;1m[#0 init_stage2(2:2)][0m: 0xc05371cd (next: 0xc1135f70)
[34;1m[#0 init_stage2(2:2)][0m: 0xc054259d (next: 0x00000000)
```

Reply:
oops! fixed